aten/src/ATen/native/BinaryOps.cpp

```diff
@@ -933,6 +934,53 @@ Tensor _test_serialization_subcmul(const Tensor& self, const Tensor& other, Scal
   return self - (other * alpha);
 }
 
+Tensor& heaviside_tensor_out(Tensor& result, const Tensor& self, const Tensor& val) {
+  TORCH_CHECK(!self.is_complex() && !result.is_complex() && !val.is_complex(),
+              "heaviside is not implemented for complex tensors.");
+
+  auto iter = TensorIterator::binary_op(result, self, val.to(self.dtype()), /*check_mem_overlap=*/true);
+  heaviside_stub(iter.device_type(), iter);
+  return result;
+}
+
+Tensor& heaviside_scalar_out(Tensor& result, const Tensor& self, Scalar val) {
+  Tensor val_t = wrapped_scalar_tensor_and_check_convert(val, self);
+  TORCH_CHECK(!self.is_complex() && !result.is_complex() && !val_t.is_complex(),
+              "heaviside is not implemented for complex tensors.");
+
+  auto iter = TensorIterator::binary_op(result, self, val_t, /*check_mem_overlap=*/true);
+  heaviside_stub(iter.device_type(), iter);
+  return result;
+}
+
+Tensor heaviside(const Tensor& self, const Tensor& val) {
+  TORCH_CHECK(!self.is_complex() && !val.is_complex(),
+    "heaviside is not implemented for complex tensors.");
+
+  Tensor result;
+  auto iter = TensorIterator::binary_op(result, self, val.to(self.dtype()));
```

Comment:
Same question about how the dtype of val is handled as above

Reply:
I'd like to keep the same dtype for self and val for the time being.